aten/src/ATen/native/vulkan/api/Context.cpp

```diff
@@ -0,0 +1,281 @@
+#include <ATen/native/vulkan/api/Context.h>
+
+#include <sstream>
+
+namespace at {
+namespace native {
+namespace vulkan {
+namespace api {
+namespace {
+
+struct Configuration final {
+#ifndef DEBUG
+  static constexpr bool kEnableValidationLayers = false;
+#else
+  static constexpr bool kEnableValidationLayers = true;
+#endif
+};
+
+VKAPI_ATTR VkBool32 VKAPI_CALL debug_report_callback_fn(
+    const VkDebugReportFlagsEXT flags,
+    const VkDebugReportObjectTypeEXT /* object_type */,
+    const uint64_t /* object */,
+    const size_t /* location */,
+    const int32_t message_code,
+    const char* const layer_prefix,
+    const char* const message,
+    void* const /* user_data */) {
+  std::stringstream stream;
+  stream << layer_prefix << " " << message_code << " " << message << std::endl;
+  const std::string log = stream.str();
+
+  if (flags & VK_DEBUG_REPORT_ERROR_BIT_EXT) {
+    LOG(ERROR) << log;
+  } else if (flags & VK_DEBUG_REPORT_WARNING_BIT_EXT) {
+    LOG(WARNING) << log;
+  } else if (flags & VK_DEBUG_REPORT_PERFORMANCE_WARNING_BIT_EXT) {
+    LOG(WARNING) << "Performance:" << log;
+  } else if (flags & VK_DEBUG_REPORT_INFORMATION_BIT_EXT) {
+    LOG(INFO) << log;
+  } else if (flags & VK_DEBUG_REPORT_DEBUG_BIT_EXT) {
+    LOG(INFO) << "Debug: " << log;
+  }
+
+  return VK_FALSE;
+}
+
+VkInstance create_instance(const bool enable_validation_layers) {
+  std::vector<const char*> enabled_instance_layers;
+  std::vector<const char*> enabled_instance_extensions;
+
+  if (enable_validation_layers) {
+    uint32_t instance_layers_count = 0;
+    VK_CHECK(vkEnumerateInstanceLayerProperties(
+        &instance_layers_count, nullptr));
+
+    std::vector<VkLayerProperties> instance_layer_properties(
+        instance_layers_count);
+
+    VK_CHECK(vkEnumerateInstanceLayerProperties(
+        &instance_layers_count,
+        instance_layer_properties.data()));
+
+    constexpr std::array<const char*, 6> requested_instance_layers{
+        "VK_LAYER_GOOGLE_unique_objects",
```

Comment:
There is some optimal validation layer order:
```
        "VK_LAYER_GOOGLE_threading",
        "VK_LAYER_LUNARG_parameter_validation",
        "VK_LAYER_LUNARG_object_tracker",
        "VK_LAYER_LUNARG_core_validation",
        "VK_LAYER_GOOGLE_unique_objects",
```
( https://www.lunarg.com/wp-content/uploads/2019/04/UberLayer_V3.pdf )

Reply:
Thanks.  I think even better is `VK_LAYER_KHRONOS_validation`:

https://vulkan.lunarg.com/doc/view/1.1.114.0/windows/khronos_validation_layer.html
https://vulkan.lunarg.com/doc/view/1.2.135.0/mac/validation_layers.html